### PR TITLE
Follow-up tasks for enabling AKS preview features

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR includes follow-up tasks for enabling AKS preview features #4617, including:

1. Operating on the hub API version instead of stable
2. Adding E2E tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: azure/services/agentpools/spec.go [hunk 1]
```diff
@@ -336,12 +335,8 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existingObj genruntime.M
...
-		hub := &asocontainerservicev1hub.ManagedClustersAgentPool{}
-		if err := agentPool.ConvertTo(hub); err != nil {
-			return nil, err
-		}
 		prev := &asocontainerservicev1preview.ManagedClustersAgentPool{}
-		if err := prev.ConvertFrom(hub); err != nil {
+		if err := prev.ConvertFrom(agentPool); err != nil {
 			return nil, err
 		}
 		if existing != nil {
```
[the remark this answers is not in the document]
Comment (willie-yao) [reply azure/services/agentpools/spec.go:342]: From what I remember, the status field was missing the new preview fields when converted from stable. I'm not sure if doing so from the hub version would fix this but I'll give it a test.

[the remark this answers is not in the document]
Comment (willie-yao) [reply azure/services/agentpools/spec.go]: Confirmed that the hack is no longer needed!

File: azure/services/agentpools/spec.go [hunk 2]
```diff
@@ -198,29 +198,26 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existingObj genruntime.M
 	defer done()
 
 	// If existing is preview, convert to stable then back to preview at the end of the function.
-	var existing *asocontainerservicev1.ManagedClustersAgentPool
-	var existingStatus asocontainerservicev1preview.ManagedClusters_AgentPool_STATUS
+	var existing *asocontainerservicev1hub.ManagedClustersAgentPool
 	if existingObj != nil {
+		hub := &asocontainerservicev1hub.ManagedClustersAgentPool{}
 		if s.Preview {
```
Comment (nojnhuh) [inline azure/services/agentpools/spec.go:204]: I think we could do something similar here to get rid of the branching:
```go
		hub := &asocontainerservicev1hub.ManagedClustersAgentPool{}
		if err := existingObj.(conversion.Convertible).ConvertTo(hub); err != nil {
			return nil, err
		}
```

File: azure/services/agentpools/spec.go [hunk 3]
```diff
@@ -198,29 +198,26 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existingObj genruntime.M
...
+		} else {
+			existingStable := existingObj.(*asocontainerservicev1.ManagedClustersAgentPool)
+			if err := existingStable.ConvertTo(hub); err != nil {
 				return nil, err
 			}
-			existing = stable
-		} else {
-			existing = existingObj.(*asocontainerservicev1.ManagedClustersAgentPool)
 		}
+		existing = hub.DeepCopy()
```
Comment (nojnhuh) [inline azure/services/agentpools/spec.go:215]: Why the deep copy here?

Comment (willie-yao) [reply azure/services/agentpools/spec.go:215]: I think this was left over from a fix I tried to do for the mismatched Status. It definitely isn't needed!

**Review by nojnhuh — COMMENTED** [review]
/lgtm
/hold for squash

**Comment by nojnhuh** [comment]
Thank you for squashing!
/hold cancel
/approve

**Pull request merged**